File: trains-gdx-core/src/main/java/com/github/caay2000/trains/Position.java
package com.github.caay2000.trains;

import com.badlogic.gdx.math.Vector2;

import java.util.Objects;
import java.util.StringJoiner;

public class Position {

    private final Vector2 vec2d;

    public Position() {
        this.vec2d = Vector2.Zero;
    }

    public Position(Vector2 vec2d) {
        this.vec2d = vec2d;
    }

    public Position(int x, int y) {
        this.vec2d = new Vector2(x, y);
    }

    public Position(float x, float y) {
        this.vec2d = new Vector2(x, y);
    }

    public Position move(Position to, float distance) {
        if (distance > vec2d.dst(to.vec2d)) {
            return new Position(to.vec2d);
        }
        Vector2 diff = copy().vec2d.sub(to.vec2d).nor();
        diff = new Vector2(diff.x * -1, diff.y * -1);
        return new Position(copy().vec2d.mulAdd(diff, distance));
    }

    public Position copy() {
        return new Position(vec2d.x, vec2d.y);
    }

    public float distanceTo(Position end) {
        return vec2d.dst(end.vec2d);
    }

    @Override
    public String toString() {
        return new StringJoiner(", ", Position.class.getSimpleName() + "[", "]")
                .add("x=" + vec2d.x)
                .add("y=" + vec2d.y)
                .toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Position position = (Position) o;
        return Objects.equals(vec2d, position.vec2d);
    }

    @Override
    public int hashCode() {
        return Objects.hash(vec2d);
    }
}

File: trains-gdx-core/src/main/kotlin/com/github/caay2000/trains/world/position/Position.kt
package com.github.caay2000.trains.world.position

import com.badlogic.gdx.math.Vector2

class Position {

    private val vec2d: Vector2

    constructor(x: Int = 0, y: Int = 0) {
        vec2d = Vector2(x.toFloat(), y.toFloat())
    }

    constructor(x: Float, y: Float) {
        vec2d = Vector2(x, y)
    }

    constructor(position: Position) {
        vec2d = Vector2(position.x, position.y)
    }

    private constructor(vec2d: Vector2) {
        this.vec2d = Vector2(vec2d.x, vec2d.y)
    }

    private fun copy(): Position =
        Position(this.vec2d)

    val x: Float get() = this.vec2d.x
    val y: Float get() = this.vec2d.y

    fun distanceTo(to: Position): Float = this.vec2d.dst(to.vec2d)

    fun move(to: Position, distance: Float) {
        if (distance <= this.distanceTo(to)) {
            val diff = this.copy().vec2d.sub(to.vec2d).nor()
            this.vec2d.mulAdd(diff.scl(-1f), distance)
        } else this.translate(to)
    }

    fun translate(position: Position) {
        this.vec2d.set(position.x, position.y)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Position
        return vec2d.x == other.vec2d.x && vec2d.y == other.vec2d.y
    }

    override fun hashCode(): Int {
        return vec2d.hashCode()
    }

    override fun toString(): String {
        return "Position(x=${vec2d.x},y=${vec2d.y})"
    }
}